libgnucash/engine/gnc-features.cpp

```diff
@@ -54,42 +53,14 @@ static QofLogModule log_module = G_LOG_DOMAIN;
 /********************************************************************\
 \********************************************************************/
 
-static void gnc_features_test_one(gpointer pkey, gpointer value,
-                  gpointer data)
-{
-    auto key = static_cast<const gchar*>(pkey);
-    auto feature_desc = static_cast<const gchar*>(value);
-    auto str = static_cast<std::string*>(data);
-
-    /* Check if this feature is in the known features list. */
-    if (features_table.find (key) != features_table.end())
-        return;
-
-    /* It is unknown, so add the description to the unknown features list: */
-    if (str->empty())
-        str->append (_("This Dataset contains features not supported "
-                       "by this version of GnuCash. You must use a "
-                       "newer version of GnuCash in order to support "
-                       "the following features:"));
-
-    str->append ("\n* ");
-    str->append (feature_desc);
-}
-
 /* Check if the session requires features unknown to this version of GnuCash.
  *
  * Returns a message to display if we found unknown features, NULL if
  * we're okay.
  */
 gchar *gnc_features_test_unknown (QofBook *book)
 {
-    std::string str;
-    auto features_used = qof_book_get_features (book);
-
-    /* Iterate over the members of this frame for unknown features */
-    g_hash_table_foreach (features_used, &gnc_features_test_one, &str);
-    g_hash_table_unref (features_used);
-    return str.empty() ? nullptr : g_strdup (str.c_str());
+    return qof_book_test_features (book, (gpointer)&features_table);
```

Comment:
Don't use `gpointer`, aka `void*`, in C++, it defeats the type system. Add this to qofbook.h:
```
#ifdef __cplusplus
using FeaturesTable = static_cast<std::unordered_map<std::string,std::string>*>(data);

char* qof_book_test_features(QofBook* book, const FeaturesTable& features_table);
bool qof_book_test_feature(QofBook* book, const char* feature);
#endif
```
You might consider having `qof_book_test_features` return a std::string and doing the `g_strdup` here.  `gnc_features_check_used` might need `return static_cast<gboolean>(qof_book_test_feature(book, feature);`  if there's a compiler warning.

Reply:
Nice, I'd thought it would be compulsory to have a `gnc-book.hpp`.

Reply:
Couldn't make it work. `gnc-book.hpp` it is.